internal/fields/validate.go

```diff
@@ -0,0 +1,196 @@
+	fis, err := ioutil.ReadDir(fieldsDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "reading directory with fields failed (path: %s)", fieldsDir)
+	}
+
+	var fields []fieldDefinition
+	for _, fi := range fis {
+		f := filepath.Join(fieldsDir, fi.Name())
+		body, err := ioutil.ReadFile(f)
+		if err != nil {
+			return nil, errors.Wrap(err, "reading fields file failed")
+		}
+
+		var u []fieldDefinition
+		err = yaml.Unmarshal(body, &u)
+		if err != nil {
+			return nil, errors.Wrap(err, "unmarshalling field body failed")
+		}
+		fields = append(fields, u...)
+	}
+	return &Validator{
+		schema: fields,
+	}, nil
+}
+
+// ValidateDocumentBody validates the provided document body.
+func (v *Validator) ValidateDocumentBody(body json.RawMessage) error {
+	var c common.MapStr
+	err := json.Unmarshal(body, &c)
+	if err != nil {
+		return errors.Wrap(err, "unmarshalling document body failed")
+	}
+	return v.validateMapElement("", c)
+}
+
+// ValidateDocumentMap validates the provided document as common.MapStr.
+func (v *Validator) ValidateDocumentMap(body common.MapStr) error {
+	return v.validateMapElement("", body)
+}
+
+func (v *Validator) validateMapElement(root string, elem common.MapStr) error {
+	var err error
+	for name, val := range elem {
+		key := strings.TrimLeft(root+"."+name, ".")
+
+		switch val.(type) {
+		case []map[string]interface{}:
+			for _, m := range val.([]map[string]interface{}) {
+				err = v.validateMapElement(key, m)
+				if err != nil {
+					return err
+				}
+			}
+		case map[string]interface{}:
+			err = v.validateMapElement(key, val.(map[string]interface{}))
+			if err != nil {
+				return err
+			}
+		default:
+			err = v.validateElementFormat(key, val)
+			if err != nil {
+				return err
+			}
+			break
+		}
+	}
+	return nil
+}
+
+func (v *Validator) validateElementFormat(key string, val interface{}) error {
+	if key == "" {
+		return nil // root key is always valid
+	}
+
+	if skipValidationForField(key) {
+		return nil // generic field, let's skip validation for now
+	}
+
+	definition := findElementDefinitionInSlice("", key, v.schema)
+	if definition == nil {
+		return fmt.Errorf(`field "%s" is not defined`, key)
+	}
+
+	err := parseElementValue(key, *definition, val)
+	if err != nil {
+		return errors.Wrap(err, "parsing field value failed")
+	}
+	return nil
+}
+
+func skipValidationForField(key string) bool {
+	return isFieldFamilyMatching("agent", key) ||
+		isFieldFamilyMatching("elastic_agent", key)
+}
+
+func isFieldFamilyMatching(family, key string) bool {
+	return key == family || strings.HasPrefix(key, family+".")
+}
+
+func findElementDefinitionInSlice(root, searchedKey string, fieldDefinitions []fieldDefinition) *fieldDefinition {
+	for _, def := range fieldDefinitions {
+		key := strings.TrimLeft(root+"."+def.Name, ".")
+		if compareKeys(key, def, searchedKey) {
+			return &def
+		}
+
+		if len(def.Fields) == 0 {
+			continue
+		}
+
+		fd := findElementDefinitionInSlice(key, searchedKey, def.Fields)
```

Comment:
To avoid star recursion here, I wonder if we should store the field definitions in a flattened manner. Then we can just iterated over the flattened list and check prefixes or do some sort of glob matching instead of recursion. Could be an optimization for a follow up PR.

Reply:
I don't remember the exact thought process now, but I had a similar discussion with myself.

 Let's consider a field in results: `apache.access.foo.bar`. If we flatten field definitions (from YML files), we'll end up with 4 records to cover the single result fields -  we need field definitions for `apache`, `apache.access`, `apache.access.foo`, `apache.access.foo.bar`.

I considered this as not a huge gain comparing to the recursive solution (more towards readability?), but I'm happy to adjust it in follow-ups.

Reply:
++ to leaving it this way for now. Only reason to try the flattening ideal would be for performance / lower memory usage but until either of those things become an issue, let's not optimize prematurely.